Introduce "priority" feature for pyobjects

terminalmage opened the pull request:
When there are multiple matches (such as when both an `os_family` and `os` grain both match the minion), the order in which they are processed is determined by the iteration order of the class' `__dict__` attribute. Since this would be different in different versions of Python, behavior would be inconsistent depending on which version of Python is being used.

This adds a `priority` attribute to the `Map` subclass which, if present, will re-order the match groups such that the results will be predictable irrespective of Python version.

terminalmage committed:
Introduce "priority" feature for pyobjects

When there are multiple matches (such as when both an os_family and os
grain both match the minion), the order in which they are processed is
determined by the iteration order of the class' __dict__ attribute.
Since this would be different in different versions of Python, behavior
would be inconsistent depending on which version of Python is being used.

This adds a priority attribute to the Map subclass which, if present,
will re-order the match groups such that the results will be predictable
irrespective of Python version.

salt-jenkins commented:
@terminalmage, thanks for your PR! By analyzing the history of the files in this pull request, we identified @borgstrom, @DmitryKuzmenko and @s0undt3ch to be potential reviewers.

terminalmage commented on tests/unit/test_pyobjects.py:
```diff
@@ -64,7 +64,7 @@
 
 class Samba(Map):
-    __merge__ = 'samba:lookup'
+    priority = ('os_family', 'os')
```
NOTE: `__merge__` is not the correct usage, `merge` would be the correct one to use. However, since we're not testing merging here, I'm simply replacing it with the new `priority` value.

terminalmage commented on tests/unit/test_pyobjects.py:
```diff
@@ -402,7 +402,6 @@ def test_requisite_implicit_list(self):
 
 
 class MapTests(RendererMixin, TestCase):
-    maxDiff = None
```
This was added earlier today for troubleshooting and is no longer necessary.

borgstrom commented on salt/utils/pyobjects.py:
```diff
@@ -320,10 +321,7 @@ def __set_attributes__(cls):
-                match = filt.__match__
-            else:
-                match = item
+            match = getattr(filt, '__match__', item)
```
+1

borgstrom commented on salt/utils/pyobjects.py:
```diff
@@ -332,6 +330,23 @@ def __set_attributes__(cls):
 
                 match_groups[grain][match][name] = filt.__dict__[name]
 
+        if hasattr(cls, 'priority'):
```
Since you changed the above line to be "EAFP", why not here too?

```
try:
    new_match_groups = OrderedDict([])
    for item in cls.priority:
        try:
            new_match_groups[item] = match_groups.pop(item)
        except AttributeError:
            log.error('pyobjects: priority must be an iterable')
        except KeyError:
            pass
    match_groups = new_match_groups
except TypeError:
    log.error('pyobjects: priority must be an iterable')
except AttrbuteError:
    pass
```

Bike shedding at this point, but figured I'd throw it out :)

terminalmage replied:
I had considered that, but I wanted to make sure the priority had an `__iter__` attribute since you can iterate over a string using a for loop. I wanted `priority = 'os_family'` to be considered invalid and to log an error.

There is definitely some room for improvement though. I just pushed another commit, let me know what you think.

terminalmage replied:
Also, I would appreciate your feedback on the documentation I added to salt/renderers/pyobjects.py. I want to make sure that what I added is accurate.

terminalmage committed:
Update test_map to test declarative ordering

terminalmage committed:
Add test for pyobjects priority

terminalmage commented:
I've rebased and updated my code to work with @DmitryKuzmenko's code from #41958. I've also update the unit tests to test the declarative ordering, and added a separate test to confirm the priority feature.

terminalmage commented:
The failing unit test is unrelated and fixed in https://github.com/saltstack/salt/pull/41999.
